content/post/2022-09-01-r-dependency/index.Rmd

```diff
@@ -0,0 +1,284 @@
+  
+  ver <- package_version(ver)
+  
+  # Here, we rely on a somewhat uncommon use of `match()`. When `match()`ing
+  # `TRUE` to something, the index of the first `TRUE` value will be returned.
+  # In other words here, we return the first R version that it superior or equal
+  # to the stated R version dependency
+  min_r_ver <- r_vers[match(TRUE, ver <= r_vers)]
+  
+  return(min_r_ver)
+  
+}, package_version("0.0.0"))
+```
+
+```{r cran-r-version-plot}
+library(ggplot2)
+
+do.call(rbind, normalised_r_deps) |> 
+  as.data.frame() |> 
+  dplyr::rename(
+    major = V1,
+    minor = V2,
+    patch = V3
+  ) |> 
+  dplyr::mutate(majorminor = paste(major, minor, sep = ".")) |> 
+  ggplot(aes(y = majorminor)) +
+    geom_bar() +
+    labs(
+      x = "Number of CRAN packages",
+      y = "Minimum required R version",
+      title = "Minimum required R version in CRAN packages"
+    ) +
+    theme_minimal()
+```
+
+The peak at R 2.10 might be related to the fact that [it is automatically added when developers embed data in their packages with `usethis::use_date()`](https://github.com/r-lib/usethis/issues/631).
+
+# How to avoid depending on a new version?
+
+For the various reasons presented above, it might not always be desirable to depend on a very recent R version.
+In this kind of situation, you may want to use the backports package.
+It reimplements many of the new features from the more recent R version.
+This way, instead of having to depend on a newer R version, you can simply add a dependency to backports, which is easier to install than a newer R version for users in highly controlled environments.
+
+Backports is not a silver bullet though, as some new features are impossible to reimplement in a package.
+Notably, this is the case of the native R pipe (`|>`), introduced in R 4.1.0.
+Roughly speaking, this is because it is not simply a new function, but rather an entire new way to read R code.
+
+# How to test you depend on the correct version?
+
+[It is easy to make a mistake when specifying a minimum R version, and to forget to you use one recent R feature](https://stat.ethz.ch/pipermail/r-package-devel/2021q1/006508.html).
+For this reason, you should always try to verify that your minimum R version claim is accurate.
+
+The most complete approach is to run your tests, or at least verify that the package can be built without errors, on all older R versions you claim to support.
+For this, you could use rig, which allows you to install multiple R version on your computer and switch between them with a single command.
+But a convenient way to do so if to rely on continuous integration platforms, where existing workflows are already set up to run on multiple R versions.
+For example, if you choose to replicate the tidyverse policy of supporting the 5 latest minor releases of R, your best bet is probably to copy the `R-CMD-check.yaml` GitHub Actions workflow from any of their repository.
```

Comment:
rather refer to https://github.com/r-lib/actions/tree/v2/examples and https://usethis.r-lib.org/reference/github_actions.html (so copy from the source not "any of their repositories")?